src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java

```diff
@@ -956,4 +959,168 @@ private Node tryFoldStringSplit(Node n, Node stringNode, Node arg1) {
+
+  /**
+   * Perform folding of chained concat functions, so `arr.concat(a).concat(b)` will be fold into
+   * `arr.concat(a,b)`;
+   */
+  private Node foldConcatChaining(ConcatFunctionCall concatFunctionCall) {
+    checkArgument(isSafeToFoldConcatChaining(concatFunctionCall), concatFunctionCall);
+
+    Node concatCall = concatFunctionCall.getCallNode();
+    Node previousConcatCall = concatFunctionCall.getCalleeNode();
+    Node arg = concatFunctionCall.getFirstArgumentNode();
+    while (arg != null) {
+      Node curentArg = arg;
+      arg = arg.getNext();
+      previousConcatCall.addChildToBack(curentArg.detach());
+    }
+    concatCall.replaceWith(previousConcatCall.detach());
+    reportChangeToEnclosingScope(previousConcatCall);
+    return previousConcatCall;
+  }
+
+  /**
+   * Check if we have this code pattern `[].concat(exactlyArrayArgument,...*)` so
+   * `removeArrayLiteralFromFrontOfConcat` can be performed.
+   */
+  private boolean isSafeToRemoveArrayLiteralFromFrontOfConcat(
+      ConcatFunctionCall concatFunctionCall) {
+    checkNotNull(concatFunctionCall);
+
+    Node arrayLiteralToRemove = concatFunctionCall.getCalleeNode();
+    if (!arrayLiteralToRemove.isArrayLit() || arrayLiteralToRemove.getChildCount() != 0) {
+      return false;
+    }
+    Node firstArg = concatFunctionCall.getFirstArgumentNode();
+    return containsExactlyArray(firstArg);
+  }
+
+  /**
+   * Check if we have this code pattern `array.concat(...*).concat(sideEffectFreeArguments)` so
+   * `foldConcatChaining` can be performed.
+   */
+  private boolean isSafeToFoldConcatChaining(ConcatFunctionCall concatFunctionCall) {
+    checkNotNull(concatFunctionCall);
+
+    Node maybeFunctionCall = concatFunctionCall.getCalleeNode();
+    if (!maybeFunctionCall.isCall()) {
+      return false;
+    }
+    ConcatFunctionCall previousConcatFunctionCall =
+        createConcatFunctionCallForNode(maybeFunctionCall);
+    if (previousConcatFunctionCall == null) {
+      return false;
+    }
+    // make sure that arguments in second concat function call can't change the array
+    // so we can fold chained concat functions
+    // to clarify, consider this code
+    // here we can't fold concatenation
+    // var a = [];
+    // a.concat(1).concat(a.push(1)); -> [1,1]
+    // a.concat(1,a.push(1)); -> [1,1,1]
+    for (Node arg = concatFunctionCall.getFirstArgumentNode(); arg != null; arg = arg.getNext()) {
+      if (mayHaveSideEffects(arg)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private abstract static class ConcatFunctionCall {
+    private final Node callNode;
+    private final Node calleeNode;
+    @Nullable
+    private final Node firstArgumentNode;
+
+    ConcatFunctionCall(Node callNode, Node calleeNode, Node firstArgumentNode) {
+      this.callNode = checkNotNull(callNode);
+      this.calleeNode = checkNotNull(calleeNode);
+      this.firstArgumentNode = firstArgumentNode;
+    }
+
+    public Node getCallNode() {
```

Comment:
Nit:

Since all the fields are private and final, there is no need to have accessor functions.
Leave them if you like, but I wouldn't bother with them.